### ISSUE TITLE
ingest/importers/regexp.py tests failing

### ISSUE THREAD

**Issue by blais** [beancount/beancount#211]
**[Original report](https://bitbucket.org/blais/beancount/issue/211) by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

```
#!python
FAIL: test_match (regexp_test.TestRegexpMixin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../beancount/utils/test_utils.py", line 224, in new_function
    return function(self, file.name)
  File ".../beancount/ingest/importers/regexp_test.py", line 45, in test_match
    self.assertTrue(importer.identify(file))
AssertionError: False is not true
```

Cause:
The tempfile is detected as "MimeType: text/x-Algol68", when the tests expect "MimeType: text/plain"

e.g.
```
#!python
>>> print(match_text)
"""
Filename: /tmp/tmptr9cn666
MimeType: text/x-Algol68
Contents:
DATE,TYPE,REF #,DESCRIPTION,FEES,AMOUNT,BALANCE
2014-04-14,BUY,14167001,BOUGHT +CSKO 50 @98.35,7.95,-4925.45,25674.63
2014-05-08,BUY,12040838,BOUGHT +HOOL 121 @79.11,7.95,-9580.26,16094.37
2014-05-11,BUY,41579908,BOUGHT +MSFX 104 @64.39,7.95,-6704.51,9389.86
2014-05-22,DIV,54857517,ORDINARY DIVIDEND~HOOL,0,28.56,9418.42
"""
```

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Very strange. Can you provide details of the 
- python interpreter
- python-magic
- filemagic
packages you've installed?

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

Sorry, didn't see your reply, this is my pip freeze:

```
beautifulsoup4==4.6.0
bottle==0.12.13
chardet==3.0.4
google-api-python-client==1.6.5
httplib2==0.10.3
lxml==4.1.1
m3-cdecimal==2.3
oauth2client==4.1.2
pluggy==0.6.0
ply==3.10
py==1.5.2
pyasn1==0.4.2
pyasn1-modules==0.2.1
pytest==3.4.0
python-dateutil==2.6.1
python-magic==0.4.15
rsa==3.4.2
six==1.11.0
uritemplate==3.0.0
```

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

Some more information... :


```
#!python
beancount/ingest/importers/regexp_test.py line 62:
    if magic:
        filetype = magic.from_file(filename, mime=True)
        if isinstance(filetype, bytes):
            print("cbytes")
            filetype = filetype.decode('utf8')
        print(filename)
        print(magic.from_file(filename))
        print(magic.Magic(mime_encoding=True).from_file(filename))
        print(magic.Magic(keep_going=True).from_file(filename))
        print(magic.Magic(keep_going=True, mime=True).from_file(filename))
        print(magic.Magic(keep_going=True, mime_encoding=True).from_file(filename))
        return filetype
```

```
/tmp/tmp855d3h7b
Algol 68 source, ASCII text
us-ascii
Algol 68 source, ASCII text\012- data
text/x-Algol68\012- application/octet-stream
\012- us-ascii
```

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

seems magic is doing something different for me

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

I tried catching the file:

```
#!bash
$ cd /tmp
$ file tmp8uq3w1jp 
tmp8uq3w1jp: Algol 68 source, ASCII text
```

Then I tried just editing it manually:

```
#!bash
$ echo 'DATE,TYPE,REF #,DESCRIPTION,FEES,AMOUNT,BALANCE
2014-04-14,BUY,14167001,BOUGHT +CSKO 50 @98.35,7.95,-4925.45,25674.63
2014-05-08,BUY,12040838,BOUGHT +HOOL 121 @79.11,7.95,-9580.26,16094.37
2014-05-11,BUY,41579908,BOUGHT +MSFX 104 @64.39,7.95,-6704.51,9389.86
2014-05-22,DIV,54857517,ORDINARY DIVIDEND~HOOL,0,28.56,9418.42' > tmp_bean
$ file tmp_bean
tmp2: Algol 68 source, ASCII text
```

Seems this is a system specific thing? I don't know anything about how file works, but if it helps:

```
#!bash
$ file -v
file-5.32
magic file from /usr/share/file/misc/magic
```

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Hmm. I don't have this pip install.
I've got "python3-magic" from Ubuntu.


bergamot:~$ python3                                                                                       
Python 3.6.3 (default, Oct 10 2017, 07:45:10)                                                             
[GCC 6.3.0 20170406] on linux                                                                             
Type "help", "copyright", "credits" or "license" for more information.                                    
>>> import magic                                                                                          
>>> magic.__file__                                                                                        
'/usr/local/lib/python3.6/site-packages/magic.py'                                                         
>>>                                                                                                       
bergamot:~$ cd /usr/local/lib/python3.6/site-packages
bergamot:/usr/.../lib/python3.6/site-packages$ cd python_
python_dateutil-2.6.0.dist-info/    python_magic-0.4.12-py3.6.egg-info/ 
bergamot:/usr/.../lib/python3.6/site-packages$ cd python_magic-0.4.12-py3.6.egg-info/
bergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ ls -l
total 20
-rw-r--r-- 1 root root   1 Jan 10  2017 dependency_links.txt
-rw-r--r-- 1 root root 104 Jan 10  2017 installed-files.txt
-rw-r--r-- 1 root root 700 Jan 10  2017 PKG-INFO
-rw-r--r-- 1 root root 222 Jan 10  2017 SOURCES.txt
-rw-r--r-- 1 root root   6 Jan 10  2017 top_level.txt
bergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ more PKG-INFO 
Metadata-Version: 1.1
Name: python-magic
Version: 0.4.12
Summary: File type identification using libmagic
Home-page: http://github.com/ahupp/python-magic
Author: Adam Hupp
Author-email: adam@hupp.org
License: MIT
Description: This module uses ctypes to access the libmagic file type
        identification library.  It makes use of the local magic database and
        supports both textual and MIME-type output.
        
Keywords: mime magic file
Platform: UNKNOWN
Classifier: Intended Audience :: Developers
Classifier: License :: OSI Approved :: MIT License
Classifier: Programming Language :: Python
Classifier: Programming Language :: Python :: 2
Classifier: Programming Language :: Python :: 3
bergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ more SSOU^C
bergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ cat SOURCES.txt 
MANIFEST.in
__init__.py
file_magic.py
magic.py
setup.cfg
setup.py
python_magic.egg-info/PKG-INFO
python_magic.egg-info/SOURCES.txt
python_magic.egg-info/dependency_links.txt
python_magic.egg-info/top_level.txt
test/test.pybergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ cat top_level.txt 
magic
bergamot:/usr/.../lib/python3.6/site-packages/python_magic-0.4.12-py3.6.egg-info$ cat installed-files.txt 
../magic.py
../__pycache__/magic.cpython-36.pyc
top_level.txt
PKG-INFO
SOURCES.txt
dependency_links.txt



You show python-magic==0.4.15 from your pip freeze.

I suspect you may have another library also called "magic"; there are several, which is why in the lib I'm trying to detect that case.
I suspect this detection is failing now.

Can you try:
python3 
>>> import magic
>>> magic.__file__

and then open that file and verify that it's the one from https://github.com/ahupp/python-magic (or not)?

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

BTW, when I use file on your example input, I get "text/plain".

bergamot:~$ file --version
file-5.32
magic file from /etc/magic:/usr/share/misc/magic

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

```
#!bash
$ python
Python 3.6.4 (default, Jan  5 2018, 02:35:40) 
[GCC 7.2.1 20171224] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import magic
>>> magic.__file__
'/<redacted>/beancount_default/.venv/lib/python3.6/site-packages/magic.py'
>>> 
$ wget https://raw.githubusercontent.com/ahupp/python-magic/master/magic.py
--2018-03-12 20:22:03--  https://raw.githubusercontent.com/ahupp/python-magic/master/magic.py
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.36.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.36.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9655 (9.4K) [text/plain]
Saving to: ‘magic.py’

magic.py                                    100%[========================================================================================>]   9.43K  --.-KB/s    in 0.001s  

2018-03-12 20:22:04 (13.1 MB/s) - ‘magic.py’ saved [9655/9655]

$ diff magic.py /<redacted>/beancount_default/.venv/lib/python3.6/site-packages/magic.py
21a22
> import os.path
65c66
< 
---
>         
78c79
<                     buf = buf.encode('utf-8', errors='replace')
---
>                    buf = buf.encode('utf-8', errors='replace')
101c102
< 
---
>         
119d119
< 
126d125
< 
139d137
< 
152a151,152
> 
> 
155,157c155
< dll = ctypes.util.find_library('magic') \
<     or ctypes.util.find_library('magic1') \
<     or ctypes.util.find_library('cygmagic-1')
---
> dll = ctypes.util.find_library('magic') or ctypes.util.find_library('magic1') or ctypes.util.find_library('cygmagic-1')
159c157
< # necessary because find_library returns None if it doesn't find the library
---
> # This is necessary because find_library returns None if it doesn't find the library
164c162
<     windows_dlls = ['magic1.dll', 'cygmagic-1.dll']
---
>     windows_dlls = ['magic1.dll','cygmagic-1.dll']
167,168c165,166
<                        # Assumes there will only be one version installed
<                        glob.glob('/usr/local/Cellar/libmagic/*/lib/libmagic.dylib'),  # flake8:noqa
---
>                          # Assumes there will only be one version installed
>                          glob.glob('/usr/local/Cellar/libmagic/*/lib/libmagic.dylib'),
171c169
<                        'linux': ['libmagic.so.1'],  # fallback for some Linuxes (e.g. Alpine) where library search does not work # flake8:noqa
---
>                        'linux': ['libmagic.so.1'],    # fallback for some Linuxes (e.g. Alpine) where library search does not work
209c207
< 
---
>     
215c213
<     # .encode('ascii').  If you use the filesystem encoding
---
>     # .encode('ascii').  If you use the filesystem encoding
```

**Comment by blais** [comment]
**Original comment by droogmic (Bitbucket: Michael Droogleever, GitHub: [droogmic](https://github.com/droogmic)).**

----------------------------------------

And outside the venv:

```
#!bash

$ python
Python 3.6.4 (default, Jan  5 2018, 02:35:40) 
[GCC 7.2.1 20171224] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import magic
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'magic'

```

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Fixed in 6c3e97015490ca353bdfd86b3725f6d74078756b
